js/create_progress_bar.js

```diff
@@ -0,0 +1,182 @@
+
+        set_bar_percentage(percentage);
+
+        if (text && text.length) {
+            messages_container.innerHTML = _.escape(text.trim()).replace(/\n/gi, '<br>');
+        }
+    }
+
+    function stop_progress_with_warning_state() {
+        is_running = false;
+
+        set_bar_percentage(null);
+        progress_bar.classList.remove('bg-info');
+        progress_bar.classList.add('bg-warning');
+    }
+
+    async function check_progress() {
+        is_running = true;
+        setTimeout(async () => {
+            try {
+                const res = await fetch(`/progress/check/${parameters.key}`, {
+                    method: 'POST',
+                    signal: abort_controller.signal,
+                });
+
+                if (res.status === 404) {
+                    const cb_err_result = parameters?.error_callback('Not found');
+                    if (!cb_err_result) {
+                        stop_progress_with_warning_state();
+                        return;
+                    }
+                }
+
+                if (res.status >= 300) {
+                    parameters?.error_callback(`Invalid response from server, expected 200 or 404, found "${res.status}".`);
+                    stop_progress_with_warning_state();
+                    return;
+                }
+
+                const json =  await res.json();
+
+                if (json['key'] && json['started_at']) {
+                    update_progress(json.current, json.max, json.data);
+                    if (
+                        (
+                            !parameters.progress_callback
+                            || (parameters.progress_callback && parameters.progress_callback(json) !== false)
+                        )
+                        && !json['finished_at']
+                    ) {
+                        // Recursive call, including the timeout
+                        await check_progress();
+                    }
+
+                    return;
+                }
+
+                parameters?.error_callback(`JSON returned by progress check endpoint is invalid.`);
+                stop_progress_with_warning_state();
+            } catch (err) {
+                parameters?.error_callback(`Request error when checking progress:\n${err.message || err.toString()}`);
+                stop_progress_with_warning_state();
+            }
+        }, start_timeout);
+    }
+
+    function error() {
+        if (!is_running) {
+            return;
+        }
+
+        try {
+            abort_controller.abort();
+        } finally {
+            abort_controller = new AbortController();
+        }
+        stop_progress_with_warning_state();
+    }
+
+    function start() {
+        progress_bar.classList.add('bg-info');
+        progress_bar.classList.remove('bg-warning');
+        set_bar_percentage(0);
+
+        check_progress().then(() => {});
+    }
+
+    return {
+        start,
+        error,
+    };
```

Comment:
It would be preferable here to return a `ProgressBar` instance that has the `start()` and `stop()` public methods.

Reply:
This means creating a class for it, which isn't the scope to me for now. This is purely internal, and since it's JS I'm still not sure about adding classes.

Reply:
We akready use modules and classes, see js/modules/Form/WebIconSelector.js

Reply:
Done